### PR TITLE
Massive conversion from Package to AutotoolsPackage

### PR DESCRIPTION
**Pull request by adamjstewart**
I keep seeing PRs that convert a single package to `AutotoolsPackage`, and I thought to myself, why not do them all at once? 😆 

This PR mostly covers the simplest examples, like all of the Xorg/X11 stuff I packaged. I didn't really go out of my way to convert more complex packages to AutotoolsPackage. Yet.

### CODE REVIEW

**Comment by tgamblin** [comment]
@adamjstewart: I think you're going to surpass @alalazo as the contribution leader for deletions. Thanks for having so much helpful OCD. 😄

**Comment by adamjstewart** [comment]
Oh don't worry, I'm just getting started 😄

File: var/spack/repos/builtin/packages/p4est/package.py [hunk 1]
```diff
@@ -42,8 +42,8 @@ class P4est(Package):
     depends_on('mpi')
     depends_on('zlib')
 
-    def install(self, spec, prefix):
-        options = [
+    def configure_args(self):
```
Comment (davydden) [inline var/spack/repos/builtin/packages/p4est/package.py]: forgot to change the class to `AutotoolsPackage`.

**Comment by citibeth** [comment]
@adamjstewart Thank you; but has anyone told you that you are a glutton for punishment?  And this comes so soon after the `PythonPackage` mega-PR.

**Comment by citibeth** [comment]
My one suggestion is... beware the *fake* Autotools packackages... that is, things that come with a script called `configure` but was NOT generated by Autotools.  IMHO, such packages should NOT subclass from `AutotoolsPackage`: they don't necessarily take the same command-line arguments, nor can they be regenerated with `autoreconf`.

Which brings us to... what do we do about the `autoreconf` stuff?  Is there a dynamic way that `AutotoolsPackage` can detect a `configure` file (or not), and run `autoreconf` if necessary?

File: var/spack/repos/builtin/packages/p4est/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/p4est/package.py]: Oops, good catch. I realize this conflicts with your #2839. Do you want me to remove this file from my changes, or do you want to close #2839? Or do you want to play chicken and see whose PR gets merged first? 😆

**Comment by adamjstewart** [comment]
> what do we do about the autoreconf stuff? Is there a dynamic way that AutotoolsPackage can detect a configure file (or not), and run autoreconf if necessary?

I'm seeing a lot of variation in the `autoreconf` methods that have been defined so far. I'm no Autotools expert, but is there any catch-all command that can be run to generate a configure script for any project? If not, it would be impossible to automate this as you suggest.

**Comment by tgamblin** [comment]
How many packages are poorly behaved autotools packages?  automating `autoreconf` is a pain, especially if you don't know what combination of autotools the package uses.

File: var/spack/repos/builtin/packages/p4est/package.py (hunk 1 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/p4est/package.py]: > Or do you want to play chicken and see whose PR gets merged first?

:+1: if your's get merged first, i will close mine :smile:

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/p4est/package.py]: Looks like you win, but GitHub isn't complaining about any conflicts? We must have identical changes.

**Comment by citibeth** [comment]
Maybe this will help clear things up?

https://devmanual.gentoo.org/general-concepts/autotools/

My guess (but it's just a guess) is the variation you see is because people
"just happened" to do it one way, and it worked.

I've always had luck just running `autoreconf`.  Here is the man page:

https://www.gnu.org/software/autoconf/manual/autoconf-2.68/html_node/autoreconf-Invocation.html

Looks like `autoreconf -ivf` will probably work in most cases.

**Comment by adamjstewart** [comment]
Ok, it looks like things are more uniform than I suspected. There aren't too many packages in Spack right now that don't come with a configure script, so it wouldn't be hard to test a major change like this. Here are our options:

1. Always run `autoreconf`. I'm worried about this, because it isn't guaranteed to always work and I don't want to test every Autotools package.
2. Only run `autoreconf` when it is defined in a package.py file. This is what we currently do, but we could try to make these methods more uniform.
3. Don't run the autoreconf phase by default. We would still define a default autoreconf method in AutotoolsPackage, but unless the user overrides `phases` in their package.py to include this phase, we don't run it. This is the option I like the best.

Another thing I can do to make our lives easier. If you run `spack create` on an AutotoolsPackage without a configure script, it doesn't realize it uses Autotools. I could change that, so that the presence of `Makefile.am`, `configure.in`, or `configure.ac` but the lack of `configure` means that we should subclass AutotoolsPackage, but add a default autoreconf method. Would that help?

**Comment by alalazo** [comment]
@adamjstewart Right now `autoreconf` does nothing. What I was thinking since some time is:
```python
if exists(configure):
    # Nothing to be done
    return

if any(x not in spec for x in ['autoconf', 'automake', 'libtool']):
    tty.error(<message that says you should depend on the packages above to run the default method>)

autoreconf('-ivf')
```

**Comment by alalazo** [comment]
> If you run spack create on an AutotoolsPackage without a configure script, it doesn't realize it uses Autotools.  I could change that, so that the presence of Makefile.am, configure.in, or configure.ac but the lack of configure means that we should subclass AutotoolsPackage, but add a default autoreconf method. Would that help?

It's not that easy: if you don't have `configure` it's very likely that you are trying to install a branch from a repository. If the same package has tarballs for, let's say, official releases you'll have cases in the same `AutotoolsPackage` subclass of versions that need `autoreconf` and versions that do not.

File: lib/spack/docs/packaging_guide.rst [hunk 2]
```diff
@@ -2007,10 +2007,15 @@ The last element of a package is its ``install()`` method.  This is
...
-.. literalinclude::  ../../../var/spack/repos/builtin/packages/mpfr/package.py
-   :pyobject: Mpfr.install
+.. code-block:: python
    :linenos:
 
+   def install(self, spec prefix):
+       configure('--prefix={0}'.format(prefix))
+
+       make()
+       make('install')
```
Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: This will just be overwritten by #2780, but I needed to do something to get the doc tests to pass.

**Comment by citibeth** [comment]
A thought experiment... what would be the consequences if we always ran
`autoreconf`?  Or at least always ran it without the `-f` (force) flag?

**Comment by alalazo** [comment]
@citibeth Just the first that comes to mind: we always need to depend on `autoconf` etc. even if that is not really needed.

**Comment by adamjstewart** [comment]
@citibeth What are the benefits of always running autoreconf?

**Comment by citibeth** [comment]
On Tue, Jan 17, 2017 at 2:39 PM, Massimiliano Culpo <
notifications@github.com> wrote:

> @citibeth <https://github.com/citibeth> Just the first that comes to
> mind: we always need to depend on autoconf etc. even if that is not
> really needed.
>

Yes, but I don't think that's a big deal:

1. SOME package will probably need it, so any significant-size build will
probably need it anyway.

 2. It's quick to install.

**Comment by tgamblin** [comment]
My issue with `autoreconf` is that it often fails, if you don't get the tool versions correctly matched to the project.  I think this will fail more often than not unless we rely on packagers to get those versions correct in dependencies.  IMHO that is a burden.

**Comment by adamjstewart** [comment]
I think it's probably best not to run `autoreconf` if we don't have to without a good reason to do so. @alalazo would you like to make the proposed changes to the `autoreconf` method in `AutotoolsPackage`? I can also do it if you want. I'll take a look at `spack create` and see if I can get it to detect whether or not a package uses Autotools even if a configure script is not present.

**Comment by alalazo** [comment]
@adamjstewart I'll try to submit a small PR with just that change within the week.

**Pull request merged**